mlx/backend/metal/allocator.cpp

```diff
@@ -158,10 +149,20 @@ void BufferCache::remove_from_list(BufferCache::BufferHolder* to_remove) {
 MetalAllocator::MetalAllocator()
     : device_(device(mlx::core::Device::gpu).mtl_device()),
       buffer_cache_(device_),
-      peak_allocated_size_(0),
       block_limit_(1.5 * device_->recommendedMaxWorkingSetSize()),
       gc_limit_(0.95 * device_->recommendedMaxWorkingSetSize()) {}
 
+size_t MetalAllocator::set_gc_limit(size_t limit) {
```

Comment:
So gc_limit of 0 would be like `set_cache_enabled(false)`?  The gc limit controls how much it keeps in the buffer cache?

Reply:
Sort of yes. But I'm having some doubts about this part of the API, I think I should remove it as it's confusing.

Let me explain what I am going for in case you have any ideas though:

The allocator has two limits:
1. The memory limit. When this limit is reached, we wait for outstanding tasks to complete before allocating any more memory.
2. The garbage collection limit. Usually the allocator works by trying to reuse from the buffer cache. If it doesn't fine a fit then we will allocate more memory (without clearing the cache). But if we are over the gc limit then we will clear memory from the cache until we are under the gc limit.

The APIs correspond exactly to those two limits but I don't think that is the right way to expose it anymore. In part because the gc limit should always be <= the memory limit (it's not really sensible otherwise) so changing the memory limit should also change the gc limit.

What you are suggesting (e.g. making the gc limit control the size of the cache) is actually a reasonable alternative. I think I like it more than what I have now.

Reply:
I think that sounds usable -- you want to tell it how much "idle" memory you can tolerate.  More might mean increased efficiency because allocating memory isn't free.  Less would mean smaller footprint, possibly at the cost of more frequent allocations.

In that case I wonder if you would want an allocation rate?  Number of buffers (and/or bytes) allocated between calls or something like that.  This would let you observe the behavior of tuning this parameter (along with the run time of your code, of course).

Reply:
For the rate, if you make it between calls you put the sampling / window on the caller.

Reply:
Along with the allocation rate, you might want the buffer pool reuse rate.

Reply:
All sounds very reasonable. I may play with that but possibly for a follow up. Prefer not to introduce too much to the API at once.